### PR TITLE
Language Server Binary Protocol Client

### PR DESCRIPTION
**Pull request by mwu-tow**
### Pull Request Description
This PR implements the binary data protocol of the Language Server.

### Important Notes
* a new RPC handler has been implemented. Currently it largely duplicates the previous json-rpc handler but it is supposed to supersede it, as it supports both json and binary rpc protocols;
* `mockall` library has been used to provide a mock client for data protocol;
* serialization code is still messy, not sure how much better can we make it, given the generated flatbuffers code;

### Checklist
Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Rust](https://github.com/luna/enso/blob/master/doc/rust-style-guide.md), [Scala](https://github.com/luna/enso/blob/master/doc/scala-style-guide.md), [Java](https://github.com/luna/enso/blob/master/doc/java-style-guide.md) or [Haskell](https://github.com/luna/enso/blob/master/doc/haskell-style-guide.md) style guides as appropriate.
- [x] All code has been tested where possible.

### CODE REVIEW

File: src/rust/ide/enso-protocol/src/handler.rs [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+                Disposition::HandleReply {id,reply} => self.process_reply(id,reply),
+                Disposition::EmitEvent {event} => self.emit_event(event),
+                Disposition::Ignore => {}
+            }
+        });
+    }
+
+    pub fn make_request<F,R>
+    (&mut self, message:&dyn IsRequest<Id=Id>, f:F) -> impl Future<Output=FallibleResult<R>>
+    where F: FnOnce(Reply) -> FallibleResult<R> {
```
Comment (farmaazon) [inline src/rust/ide/enso-protocol/src/handler.rs:134]: Can we name it `send_request`? Do it send message immediately or shall I poll the returned future first? It should be covered in docs. And what is `f` here? And why we have &dyn? 

And I was reading this file from top to bottom, and missed the actual docs for this function in `Handler`, so I would put here just a note _See `Handler::make_request` docs_

File: src/rust/ide/enso-protocol/src/lib.rs [hunk 2]
```diff
@@ -1,6 +1,10 @@
 //! Client side implementation of Enso protocol.
 
+#![feature(type_alias_impl_trait)]
+#![feature(trait_alias)]
+//#![feature(async_closure)]
```
Comment (farmaazon) [inline src/rust/ide/enso-protocol/src/lib.rs:5]: commented code

File: src/rust/ide/src/controller/module.rs [hunk 3]
```diff
@@ -116,9 +116,13 @@ impl Handle {
...
-    (path:Path, model:Rc<model::Module>, language_server:Rc<language_server::Connection>, parser:Parser)
-    -> Self {
-        let logger = Logger::new(format!("Module Controller {}", path));
+    ( parent          : Logger
+    , path            : Path
+    , model           : Rc<model::Module>
+    , language_server : Rc<language_server::Connection>
+    , parser          : Parser
+    ) -> Self {
+        let logger = parent.sub(format!("Module Controller {}", path));
```
Comment (farmaazon) [inline src/rust/ide/src/controller/module.rs:125]: Same questions as for `controller::Graph`

File: src/rust/ide/src/controller/project.rs [hunk 4]
```diff
@@ -104,18 +111,20 @@ mod test {
...
-            let connection     = language_server::Connection::new_mock(client);
-            let project        = controller::Project::new(connection);
-            let module         = project.module_controller(path.clone()).await.unwrap();
-            let same_module    = project.module_controller(path.clone()).await.unwrap();
-            let another_module = project.module_controller(another_path.clone()).await.unwrap();
+            json_client.set_file_read_result(file_path, Ok(response::Read{contents}));
+            let json_connection   = language_server::Connection::new_mock(json_client);
+            let binary_connection = binary::Connection::new_mock(default());
+            let project           = controller::Project::new(default(),json_connection,
+                                                             binary_connection);
```
Comment (farmaazon) [inline src/rust/ide/src/controller/project.rs:124]: We usually don't align to the (, but just make a single ident.

File: src/rust/ide/src/controller/graph.rs [hunk 5]
```diff
@@ -395,16 +395,17 @@ pub struct Handle {
 impl Handle {
 
     /// Creates a new controller. Does not check if id is valid.
-    pub fn new_unchecked(module:Rc<model::Module>, parser:Parser, id:Id) -> Handle {
+    pub fn new_unchecked(parent:Logger, module:Rc<model::Module>, parser:Parser, id:Id) -> Handle {
```
Comment (farmaazon) [inline src/rust/ide/src/controller/graph.rs:398]: Take the logger by ref. Then we avoid many clone_refs in our code

File: src/rust/ide/src/controller/project.rs [hunk 6]
```diff
@@ -23,19 +24,25 @@ type ModulePath = controller::module::Path;
...
     pub module_registry     : Rc<model::module::registry::Registry>,
     pub parser              : Parser,
     pub logger              : Logger,
 }
 
 impl Handle {
     /// Create a new project controller.
-    pub fn new(language_server_client:language_server::Connection) -> Self {
+    pub fn new
+    ( parent                 : Logger
```
Comment (farmaazon) [inline src/rust/ide/src/controller/project.rs:36]: Here also we could take by ref.

File: src/rust/ide/enso-protocol/Cargo.toml [hunk 7]
```diff
@@ -11,13 +11,32 @@ crate-type = ["cdylib", "rlib"]
...
-hex        = { version = "0.4.2"                                             }
-serde      = { version = "1.0"  , features = ["derive"]                      }
-serde_json = { version = "1.0"                                               }
-sha3       = { version = "0.8.2"                                             }
-uuid       = { version = "0.8"  , features = ["serde", "v4", "wasm-bindgen"] }
+chrono      = { version = "0.4"  , features = ["serde"]                       }
+failure     = { version = "0.1.8"                                             }
+flatbuffers = { version = "0.5"                                               }
+futures     = { version = "0.3.1"                                             }
+mockall = { version = "0.7.1", features = ["nightly"] }
```
Comment (wdanilo) [inline src/rust/ide/enso-protocol/Cargo.toml:21]: spacing

File: src/rust/ide/enso-protocol/src/binary/client.rs [hunk 8]
```diff
@@ -0,0 +1,369 @@
...
+    /// handled. Returns a function so that it may be passed to the `Handler`.
+    fn processor
+    (logger:Logger)
+    -> impl FnMut(TransportEvent) -> Disposition<Uuid, FromServerPayloadOwned,Notification> + 'static {
+        move |event:TransportEvent| {
+            let binary_data = match event {
+                TransportEvent::BinaryMessage(data) =>
+                    data,
+                _ =>
+                    return Disposition::error(UnexpectedTextMessage),
```
Comment (wdanilo) [inline src/rust/ide/enso-protocol/src/binary/client.rs:111]: make it one liners pls without `=>` alignment

**Pull request merged**